File: 0009-palindrome-number/0009-palindrome-number.cpp
class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0 || (x % 10 == 0 && x != 0)) 
            return false;

        int reversed = 0;
        int org = x;

        while (x != 0) {
            int digit = x % 10;
            x /= 10;

            // Check for overflow before updating reversed
            if (reversed > (INT_MAX / 10) || (reversed == INT_MAX / 10 && digit > 7)) 
                return false; // Overflow would occur
            
            reversed = reversed * 10 + digit;
        }

        return org == reversed;
    }
};

File: 0009-palindrome-number/0009-palindrome-number.js
/**
 * @param {number} x
 * @return {boolean}
 */
var isPalindrome = function(x) {
    if (x < 0 || (x % 10 === 0 && x !== 0)) {
        return false;
    }

    let reversed = 0;
    let original = x;

    while (x > 0) {
        let digit = x % 10;
        x = Math.floor(x / 10);

        // Check for overflow before updating reversed
        if (reversed > (Number.MAX_SAFE_INTEGER / 10)) {
            return false; // Overflow would occur
        }

        reversed = reversed * 10 + digit;
    }

    return original === reversed;
};
